rats-processors/src/python/rats/pipelines_app/_decorators.py

```diff
@@ -0,0 +1,61 @@
+from collections.abc import Callable
+from typing import Concatenate, ParamSpec, TypeVar
+
+from rats import apps
+from rats.processors import ux
+
+T_ServiceType = TypeVar("T_ServiceType")
+P = ParamSpec("P")
+T_Container = TypeVar("T_Container", bound=apps.Container)
+
+
+def _process_method_to_task_method(
+    process_method: Callable[Concatenate[T_Container, P], ux.T_Processor_Output],
+) -> Callable[[T_Container], ux.UPipeline]:
+    """Convert a process method to a method returning a task wrapping over the process method."""
+
+    class _Processor(ux.IProcess):
+        process = process_method
+
+    _Processor.__name__ = process_method.__name__
+
+    def get_task(self: T_Container) -> ux.UPipeline:
+        return ux.UTask(_Processor)
+
+    get_task.__name__ = process_method.__name__
+
+    return get_task
+
+
+def task(
+    process_method: Callable[Concatenate[T_Container, P], ux.T_Processor_Output],
+) -> Callable[[T_Container], ux.UPipeline]:
+    """Decorator creating a pipeline service from a process method.
+
+    The pipeline will be a task wrapping over the method.
+
+    The name of the pipeline will be name of the method.
+    """
+    task_method = _process_method_to_task_method(process_method)
+    return apps.autoid_service(task_method)
+
+
+def pipeline(
+    get_pipeline_method: Callable[[T_Container], ux.Pipeline[ux.TInputs, ux.TOutputs]],
+) -> Callable[[T_Container], ux.Pipeline[ux.TInputs, ux.TOutputs]]:
+    """Decorator creating a pipeline service.
+
+    The name of the pipeline will be name of the method.
+    """
+    name = get_pipeline_method.__name__
+
+    def get_pipeline_and_rename(
+        self: T_Container,
```

Comment:
The decorator takes an unbound method and returns an unbound method of the exact same type and signature.
The type of the input method includes the type of the Container it is defined on.  If I used Container rather than T_Container the type of the returned method would be different.

As for renaming `self` to `container` - I prefer self because it makes it clear that the intent here is to wrap over a method.  It does not really matter to the outside because the signature is copied from the input method using the wraps decorator, but I think it's clearer.

Reply:
Ok, I agree with the TypeVar call.

Self is confusing to me because I think it should not reflect intention but what it actually is in the context is used... But your decision